src/file_analysis/analyzer/x509/functions.bif

```diff
@@ -924,3 +972,125 @@ function x509_set_certificate_cache_hit_callback%(f: string_any_file_hook%) : bo
 
 	return zeek::val_mgr->True();
 	%}
+
+## This function checks a hostname against the name given in a certificate subject/SAN, including
+## our interpretation of RFC6128 wildcard expansions. This specifically means that wildcards are
+## only allowed in the leftmost label, wildcards only span one label, the wildcard has to be the
+## last character before the label-separator, but additional characters are allowed before it, and
+## the wildcard has to be at least at the third level (so *.a.b).
+##
+## hostname: Hostname to test
+##
+## certname: Name given in the CN/SAN of a certificate; wildcards will be expanded
+##
+## Returns: True if the hostname matches.
+##
+## .. zeek:see:: x509_check_cert_hostname
+function x509_check_hostname%(hostname: string, certname: string%): bool
+	%{
+	if ( check_hostname(hostname->ToStdStringView(), certname->ToStdStringView()) )
+		return zeek::val_mgr->True();
+
+	return zeek::val_mgr->False();
+	%}
+
+## This function checks if a hostname matches one of the hostnames given in the certificate.
+##
+## For our matching we adhere to RFC6128 for the labels (see :zeek:id:`x509_check_hostname`).
+## Furthermore we adhere to RFC2818 and check only the names given in the SAN, if a SAN is present,
+## ignoring CNs in the Subject. If no SAN is present, we will use the last CN in the subject
+## for our tests.
+##
+## cert: The X509 certificate opaque handle.
+##
+## hostname: Hostname to check
+##
+## Returns: empty string if the hostname does not match; matched name (which can contain wildcards)
+##          if it did.
+##
+## .. zeek:see:: x509_check_hostname
+function x509_check_cert_hostname%(cert_opaque: opaque of x509, hostname: string%): string
+	%{
+	auto* cert_handle = (zeek::file_analysis::detail::X509Val *) cert_opaque;
+	std::string_view hostview = hostname->ToStdStringView();
+
+	X509* cert = cert_handle->GetCertificate();
+	if ( ! cert )
+		{
+		zeek::emit_builtin_error(zeek::util::fmt("No certificate in opaque"));
+		return zeek::make_intrusive<zeek::StringVal>("");
+		}
+
+	// According to RFC5280 (4.2.1.6) and RFC2818 (3.1), if the SAN is present, the subject
+	// of the certificate is ignored. Let's start by looking at the SAN.
+	auto sanpos = X509_get_ext_by_NID(cert, NID_subject_alt_name, -1);
```

Comment:
Just checking because I don't know the OpenSSL APIs: I assume there's nothing that could leak memory in the calls below?

Reply:
I *hope* not. Our leak checker also checks this codepath (and found one case that I missed at first). I went through it again afterwards and I think it is fine now.